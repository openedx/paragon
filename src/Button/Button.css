/**
 * Do not edit directly
 * Generated on Wed, 01 Mar 2023 00:27:19 GMT
 */

:root {
  --pgn-transition-btn: none;
  --pgn-spacing-btn-focus-gap: 1px;
  --pgn-spacing-btn-block-spacing-y: .5rem;
  --pgn-size-btn-focus-width: 2px;
  --pgn-typography-btn-line-height-sm: var(--pgn-typography-input-btn-line-height-sm);
  --pgn-typography-btn-line-height-base: var(--pgn-typography-input-btn-line-height-base);
  --pgn-typography-btn-font-weight: var(--pgn-typography-font-weight-normal);
  --pgn-typography-btn-font-size-lg: var(--pgn-typography-input-btn-font-size-lg);
  --pgn-typography-btn-font-size-sm: var(--pgn-typography-input-btn-font-size-sm);
  --pgn-typography-btn-font-size-base: var(--pgn-typography-input-btn-font-size-base);
  --pgn-typography-btn-font-family: var(--pgn-typography-input-btn-font-family);
  --pgn-spacing-btn-focus-border-gap: calc(var(--pgn-size-btn-focus-width) + var(--pgn-spacing-btn-focus-gap));
  --pgn-spacing-btn-padding-x-sm: var(--pgn-spacing-input-btn-padding-sm-x);
  --pgn-spacing-btn-padding-x-lg: var(--pgn-spacing-input-btn-padding-lg-x);
  --pgn-spacing-btn-padding-x-base: var(--pgn-spacing-input-btn-padding-x);
  --pgn-spacing-btn-padding-y-sm: var(--pgn-spacing-input-btn-padding-sm-y);
  --pgn-spacing-btn-padding-y-lg: var(--pgn-spacing-input-btn-padding-lg-y);
  --pgn-spacing-btn-padding-y-base: var(--pgn-spacing-input-btn-padding-y);
  --pgn-size-btn-border-radius-sm: var(--pgn-size-border-radius-sm);
  --pgn-size-btn-border-radius-lg: var(--pgn-size-border-radius-lg);
  --pgn-size-btn-border-radius-base: var(--pgn-size-border-radius-base);
  --pgn-typography-btn-line-height-lg: var(--pgn-typography-input-btn-line-height-lg);
  --pgn-spacing-btn-focus-distance-to-border: calc(var(--pgn-spacing-btn-focus-border-gap) + var(--pgn-size-btn-border-width));
  --pgn-size-btn-focus-border-radius: calc(var(--pgn-size-btn-border-radius-base) + var(--pgn-spacing-btn-focus-border-gap));
  --pgn-size-btn-border-width: var(--pgn-size-input-btn-border-width);
}
