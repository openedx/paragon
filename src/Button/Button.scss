@import "variables";
@import "~bootstrap/scss/button-group";

@mixin button-variant(
  $background, $border, $hover-background: darken($background, 7.5%),
  $hover-border: darken($border, 10%), $active-background: darken($background, 10%),
  $active-border: darken($border, 12.5%), $color: color-yiq($background), $hover-color: color-yiq($hover-background),
  $active-color: color-yiq($active-background)
) {
  color: $color;

  @include gradient-bg($background);

  border-color: $border;

  @include box-shadow($btn-box-shadow);

  @include hover {
    color: $hover-color;

    @include gradient-bg($hover-background);

    border-color: $hover-border;
  }

  // Disabled comes first so active can properly restyle
  &.disabled,
  &:disabled {
    color: $color;
    background-color: $background;
    border-color: $border;

    // Remove CSS gradients if they're enabled
    @if $enable-gradients {
      background-image: none;
    }
  }

  &:not(:disabled):not(.disabled):active,
  &:not(:disabled):not(.disabled).active,
  .show > &.dropdown-toggle {
    color: $active-color;
    background-color: $active-background;

    @if $enable-gradients {
      background-image: none; // Remove the gradient for the pressed/active state
    }

    border-color: $active-border;
  }
}

// OVERRIDE FROM BOOTSTRAP
// We do this to better control active and focus states.
@mixin button-outline-variant(
  $color, $color-hover: color-yiq($color), $active-background: $color, $active-border: $color
) {
  color: $color;
  border-color: $color;

  @include hover {
    color: $color-hover;
    background-color: $active-background;
    border-color: $active-border;
  }

  &.disabled,
  &:disabled {
    color: $color;
    background-color: transparent;
    border-color: $color;
  }

  &:not(:disabled):not(.disabled):active,
  &:not(:disabled):not(.disabled).active,
  .show > &.dropdown-toggle {
    color: color-yiq($active-background);
    background-color: $active-background;
    border-color: $active-border;
  }
}

// OVERRIDE FROM BOOTSTRAP
// No changes made. We do this to keep all button related mixins together,
//
// Button sizes
@mixin button-size($padding-y, $padding-x, $font-size, $line-height, $border-radius) {
  padding: $padding-y $padding-x;

  @include font-size($font-size);

  line-height: $line-height;

  // Manually declare to provide an override to the browser default
  @include border-radius($border-radius, 0);
}

// ------------------------
// BELOW ARE EDX EXTENSIONS
// ------------------------

//
// A new variant of button (inverse)
//
@mixin button-inverse-variant(
  $color, $background: color-yiq($color), $active-background: $color, $active-border: $color
) {
  color: $color;
  border-color: transparent;
  background-color: $background;

  &:not(:disabled):not(.disabled) {
    @include hover {
      color: darken($color, 7.5%);
      background-color: darken($background, 7.5%);
      border-color: transparent;
    }
  }

  &.disabled,
  &:disabled {
    color: $color;
    background-color: $background;
  }

  &:not(:disabled):not(.disabled):active,
  &:not(:disabled):not(.disabled).active,
  .show > &.dropdown-toggle {
    color: darken($color, 10%);
    background: #EEEEEE;
  }
}

@mixin button-focus($ring-color) {
  &.focus,
  &:focus {
    position: relative;
    outline: 0;
    box-shadow: none;

    &::before {
      content: "";
      position: absolute;
      top: -1 * $btn-focus-distance-to-border;
      right: -1 * $btn-focus-distance-to-border;
      bottom: -1 * $btn-focus-distance-to-border;
      left: -1 * $btn-focus-distance-to-border;
      border: solid $btn-focus-width $ring-color;
      border-radius: $btn-focus-border-radius;
    }

    &.btn-lg::before {
      border-radius: $btn-focus-border-radius;
    }

    &.btn-sm::before {
      border-radius: $btn-focus-border-radius;
    }

    &:active,
    &.active {
      &::before {
        opacity: .75;
      }
    }

    // Disabled buttons should not have focus rings
    &:disabled,
    &.disabled {
      &::before {
        display: none;
      }
    }
  }
}

//
// Override Bootstrap's button definition.
// We do this to get more control over focus
// and hover states, as well as to define
// an extra button variant: btn-inverse-${color}
//

//
// Base styles
//

.btn {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  font-family: $btn-font-family;
  font-weight: $btn-font-weight;
  color: $body-color;
  text-align: center;
  vertical-align: middle;
  user-select: none;
  background-color: transparent;
  border: $btn-border-width solid transparent;

  @include button-size($btn-padding-y, $btn-padding-x, $btn-font-size, $btn-line-height, $btn-border-radius);
  @include transition($btn-transition);

  @include hover {
    color: $body-color;
    text-decoration: none;
  }

  // Disabled comes first so active can properly restyle
  &.disabled,
  &:disabled {
    opacity: $btn-disabled-opacity;
  }

  // Button Icon Sizes

  .btn-icon-before {
    margin-inline-end: .5rem;
    margin-left: -.25em;

    [dir="rtl"] & {
      margin-right: -.25em;
      margin-left: .5rem;
    }
  }

  .btn-icon-after {
    margin-inline-start: .5rem;
    margin-right: -.25em;

    [dir="rtl"] & {
      margin-right: .5rem;
      margin-left: -.25em;
    }
  }
}

// Future-proof disabling of clicks on `<a>` elements
a.btn.disabled,
fieldset:disabled a.btn {
  pointer-events: none;
}

//
// Alternate buttons
//

@each $color, $value in $theme-colors {
  .btn-#{$color} {
    $background: $value;
    $border: $value;

    $hover-background: theme-color($color, "hover");
    $hover-border: theme-color($color, "hover");
    $active-background: theme-color($color, "active");
    $active-border: theme-color($color, "active");

    // Defaults
    $hover-background: darken($background, 7.5%) !default;
    $hover-border: darken($border, 10%) !default;
    $active-background: darken($background, 10%) !default;
    $active-border: darken($border, 12.5%) !default;

    @include button-variant(
      $background,
      $border,
      $hover-background,
      $hover-border,
      $active-background,
      $active-border
    );
    @include button-focus(theme-color($color, "focus"));
  }

  .btn-outline-#{$color} {
    $text-and-border: $value;

    $color-hover: theme-color($color, "hover");
    $active-background: theme-color($color, "background");
    $active-border: theme-color($color, "active");

    // Defaults
    $color-hover: color-yiq($value) !default;
    $active-background: $value !default;
    $active-border: $value !default;

    @include button-outline-variant(
      $text-and-border,
      $color-hover,
      $active-background,
      $active-border,
    );
    @include button-focus(theme-color($color, "focus"));
  }

  .btn-inverse-#{$color} {
    $text: $value;

    $active-background: theme-color($color, "background");
    $active-border: theme-color($color, "active");

    // Defaults
    $background: color-yiq($value) !default;
    $active-background: $value !default;
    $active-border: $value !default;

    @include button-inverse-variant(
      $text,
      $background,
      $active-background,
      $active-border,
    );
    @include button-focus($white);
  }

  .btn-inverse-outline-#{$color} {
    $text-and-border: $white;

    $color-hover: theme-color($color, "hover");
    $active-background: theme-color($color, "background");
    $active-border: transparent;

    // Defaults
    $color-hover: color-yiq($value) !default;
    $active-background: $value !default;
    $active-border: $value !default;

    @include button-outline-variant(
      $text-and-border,
      $color-hover,
      $active-background,
      $active-border,
    );
    @include button-focus($white);
  }
}

//
// Tertiary buttons
//

.btn-tertiary {
  $background: $btn-tertiary-bg;
  $border: transparent;
  $hover-background: $btn-tertiary-hover-bg;
  $hover-border: transparent;
  $active-background: $btn-tertiary-active-bg;
  $active-border: transparent;

  @include button-variant(
    $background,
    $border,
    $hover-background,
    $hover-border,
    $active-background,
    $active-border,
    $btn-tertiary-color,
    $btn-tertiary-color,
    $btn-tertiary-color
  );
  @include button-focus(theme-color("primary", "focus"));
}

.btn-inverse-tertiary {
  $background: $btn-inverse-tertiary-bg;
  $border: transparent;
  $hover-background: $btn-inverse-tertiary-hover-bg;
  $hover-border: transparent;
  $active-background: $btn-inverse-tertiary-active-bg;
  $active-border: transparent;

  @include button-variant(
    $background,
    $border,
    $hover-background,
    $hover-border,
    $active-background,
    $active-border,
    $btn-inverse-tertiary-color,
    $btn-inverse-tertiary-color,
    $btn-inverse-tertiary-color
  );
  @include button-focus($white);
}

//
// Link buttons
//

// Make a button look and behave like a link
.btn-link {
  font-weight: $font-weight-normal;
  color: $link-color;
  text-decoration: $link-decoration;

  @include hover {
    color: $link-hover-color;
    text-decoration: $link-hover-decoration;
  }

  &:focus,
  &.focus {
    text-decoration: $link-hover-decoration;
    box-shadow: none;
  }

  &:disabled,
  &.disabled {
    color: $btn-link-disabled-color;
    pointer-events: none;
  }

  // No need for an active state here
}

.btn-light,
.btn-dark {
  @include button-focus(theme-color("primary", "focus"));
}

//
// Button Sizes
//

.btn-lg {
  @include button-size(
    $btn-padding-y-lg, $btn-padding-x-lg, $btn-font-size-lg, $btn-line-height-lg, $btn-border-radius-lg
  );
}

.btn-sm {
  @include button-size(
    $btn-padding-y-sm, $btn-padding-x-sm, $btn-font-size-sm, $btn-line-height-sm, $btn-border-radius-sm
  );
}

//
// Block button
//

.btn-block {
  display: flex;
  width: 100%;

  // Vertically space out multiple block buttons
  + .btn-block {
    margin-top: $btn-block-spacing-y;
  }
}

.btn-inline {
  line-height: calc(#{$line-height-base}em - #{2 * $btn-border-width});
  font-size: inherit;
  vertical-align: bottom;
  padding: 0 .25em;
}

// Specificity overrides
input[type="submit"],
input[type="reset"],
input[type="button"] {
  &.btn-block {
    width: 100%;
  }
}

[dir="rtl"] .btn-group > .btn:not(:last-child):not(.dropdown-toggle),
[dir="rtl"] .btn-group > .btn-group:not(:last-child) > .btn {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

[dir="rtl"] .btn-group > .btn:not(:first-child),
[dir="rtl"] .btn-group > .btn-group:not(:first-child) > .btn {
  border-radius: $btn-border-radius 0 0 $btn-border-radius;
}

[dir="rtl"] .btn-group > .btn:first-child,
[dir="rtl"] .btn-group > .btn-group:first-child > .btn {
  border-top-right-radius: $btn-border-radius;
  border-bottom-right-radius: $btn-border-radius;
}
