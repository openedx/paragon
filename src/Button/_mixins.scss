// OVERRIDE FROM BOOTSTRAP
// No changes made. We do this to keep all button related mixins together,
//
// Button sizes
@mixin button-size($padding-y, $padding-x, $font-size, $line-height, $border-radius) {
  padding: $padding-y $padding-x;

  @include font-size($font-size);

  line-height: $line-height;

  // Manually declare to provide an override to the browser default
  @include border-radius($border-radius, 0);
}
