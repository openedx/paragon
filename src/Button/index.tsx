import React from 'react';
import PropTypes, { type Requireable } from 'prop-types';
import classNames from 'classnames';
import BaseButton, { type ButtonProps as BaseButtonProps } from 'react-bootstrap/Button';
import BaseButtonGroup, { type ButtonGroupProps as BaseButtonGroupProps } from 'react-bootstrap/ButtonGroup';
import BaseButtonToolbar, { type ButtonToolbarProps } from 'react-bootstrap/ButtonToolbar';
import type { ComponentWithAsProp } from '../utils/types/bootstrap';

import Icon from '../Icon';

interface ButtonProps extends Omit<BaseButtonProps, 'size'> {
  /**
   * An icon component to render. Example:
   * ```
   * import { Close } from '@openedx/paragon/icons';
   * <Button iconBefore={Close}>Close</Button>
   * ```
   */
  iconBefore?: React.ComponentType;
  /**
   * An icon component to render. Example:
   * ```
   * import { Close } from '@openedx/paragon/icons';
   * <Button iconAfter={Close}>Close</Button>
   * ```
   */
  iconAfter?: React.ComponentType;
  size?: 'sm' | 'md' | 'lg' | 'inline';
}

type ButtonType = ComponentWithAsProp<'button', ButtonProps> & { Deprecated?: any };

const Button: ButtonType = React.forwardRef<HTMLButtonElement, ButtonProps>(({
  children,
  iconAfter,
  iconBefore,
  size,
  ...props
}, ref) => (
  <BaseButton
    size={size as 'sm' | 'lg' | undefined} // Bootstrap's <Button> types do not allow 'md' or 'inline', but we do.
    {...props}
    className={classNames(props.className)}
    ref={ref}
  >
    {iconBefore && <Icon className="btn-icon-before" size={size} src={iconBefore} />}
    {children}
    {iconAfter && <Icon className="btn-icon-after" size={size} src={iconAfter} />}
  </BaseButton>
));

Button.propTypes = {
  /** Specifies class name to apply to the button */
  className: PropTypes.string,
  /** Disables the Button, preventing mouse events, even if the underlying component is an `<a>` element */
  disabled: PropTypes.bool,
  /** Specifies the text that is displayed within the button. */
  children: PropTypes.node.isRequired,
  /** A function that would specify what the button should do when the `onClick` event is triggered.
   * For example, the button could launch a `Modal`. The default is an empty function. */
  onClick: PropTypes.func,
  /** A function that would specify what the button should do when the `onKeyDown` event is triggered.
   * For example, this could handle using the `Escape` key to trigger the button's action.
   * The default is an empty function. */
  onKeyDown: PropTypes.func,
  /** Used to set the `type` attribute on the `button` tag.  The default type is `button`. */
  type: PropTypes.string,
  /** Specifies variant to use.
   * Can be on of the base variants: `primary`, `secondary`, `success`, `danger`, `warning`, `info`, `dark`,
   * `light`, `link`
   *
   * as well as one of the customized variants (= base variant prefixed with `inverse-`, `outline-`
   * or `inverse-outline-`)
   * */
  variant: PropTypes.string,
  /** An icon component to render.
   * Example import of a Paragon icon component: `import { Check } from '@openedx/paragon/icons';` */
  iconBefore: PropTypes.elementType as Requireable<React.ComponentType>,
  /** An icon component to render.
   * Example import of a Paragon icon component: `import { Check } from '@openedx/paragon/icons';` */
  iconAfter: PropTypes.elementType as Requireable<React.ComponentType>,
  // The 'as' type casting above is required for TypeScript checking, because the 'PropTypes.elementType' type normally
  // allows strings as a value (for use cases like 'div') but we don't support that for <Icon />/iconBefore/iconAfter.
  // The React TypeScript type definitions are more specific (React.ComponentType vs React.ElementType).
};

Button.defaultProps = {
  ...Button.defaultProps,
  children: undefined,
  className: undefined,
  iconBefore: undefined,
  iconAfter: undefined,
  disabled: false,
};

// We could just re-export 'ButtonGroup' and 'ButtonToolbar', but we currently
// override them to add propTypes validation at runtime, since most Paragon
// consumers aren't using TypeScript yet. We also force ButtonGroup's 'size'
// prop to accept our custom values of 'md' and 'inline' which are used in
// Paragon but not used in the base Bootstrap classes.

interface ButtonGroupProps extends Omit<BaseButtonGroupProps, 'size'> {
  size?: 'sm' | 'md' | 'lg' | 'inline';
}

const ButtonGroup: ComponentWithAsProp<'div', ButtonGroupProps> = (
  React.forwardRef<HTMLButtonElement, ButtonGroupProps>(({ size, ...props }, ref) => (
    <BaseButtonGroup size={size as 'sm' | 'lg'} {...props} ref={ref} />
  ))
);

ButtonGroup.propTypes = {
  /** Specifies element type for this component. */
  as: PropTypes.elementType,
  /** An ARIA role describing the button group. */
  role: PropTypes.string,
  /** Specifies the size for all Buttons in the group. */
  size: PropTypes.oneOf(['sm', 'md', 'lg', 'inline']),
  /** Display as a button toggle group. */
  toggle: PropTypes.bool,
  /** Specifies if the set of Buttons should appear vertically stacked. */
  vertical: PropTypes.bool,
  /** Overrides underlying component base CSS class name */
  bsPrefix: PropTypes.string,
};

ButtonGroup.defaultProps = {
  as: 'div',
  role: 'group',
  toggle: false,
  vertical: false,
  bsPrefix: 'btn-group',
  size: 'md',
};

const ButtonToolbar: ComponentWithAsProp<'div', ButtonToolbarProps> = (
  React.forwardRef<HTMLButtonElement, ButtonToolbarProps>((props, ref) => (
    <BaseButtonToolbar {...props} ref={ref} />
  ))
);

ButtonToolbar.propTypes = {
  /** An ARIA role describing the button group. */
  role: PropTypes.string,
  /** Overrides underlying component base CSS class name */
  bsPrefix: PropTypes.string,
};

ButtonToolbar.defaultProps = {
  role: 'toolbar',
  bsPrefix: 'btn-toolbar',
};

export default Button;
export { ButtonGroup, ButtonToolbar };
