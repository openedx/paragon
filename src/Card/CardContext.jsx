import React, { createContext } from 'react';
import PropTypes from 'prop-types';

const CardContext = createContext({});

function CardContextProvider({
  orientation,
  children,
  isLoading,
  variant,
}) {
  return (
    <CardContext.Provider value={{ orientation, isLoading, variant }}>
      {children}
    </CardContext.Provider>
  );
}

CardContextProvider.propTypes = {
  /** Specifies which orientation to use. */
  orientation: PropTypes.oneOf(['horizontal', 'vertical']),
  /** Specifies loading state. */
  isLoading: PropTypes.bool,
  /** Specifies content of the component. */
  children: PropTypes.node,
  /** Specifies `Card` style variant */
  variant: PropTypes.oneOf(['light', 'dark', 'muted']),
};

CardContextProvider.defaultProps = {
  orientation: 'vertical',
  isLoading: false,
  children: null,
  variant: 'light',
};

export { CardContextProvider };
export default CardContext;
