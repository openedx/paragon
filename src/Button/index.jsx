import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Button from 'react-bootstrap/Button';
import BaseButtonGroup from 'react-bootstrap/ButtonGroup';
import BaseButtonToolbar from 'react-bootstrap/ButtonToolbar';
import ButtonDeprecated from './deprecated';

import Icon from '../Icon';

const WrappedButton = React.forwardRef(({
  children,
  iconAfter,
  iconBefore,
  ...props
}, ref) => (
  <Button
    {...props}
    className={classNames(props.className)}
    ref={ref}
  >
    {iconBefore && <Icon className="btn-icon-before" size={props.size} src={iconBefore} />}
    {children}
    {iconAfter && <Icon className="btn-icon-after" size={props.size} src={iconAfter} />}
  </Button>
));

WrappedButton.propTypes = {
  ...Button.propTypes,
  /** Docstring for the children prop */
  children: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  /** Docstring for className... A class name to append to the button */
  className: PropTypes.string,
  /** An icon component to render.
   * Example import of a Paragon icon component: `import { Check } from '@edx/paragon/icons';` */
  iconBefore: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  /** An icon component to render.
   * Example import of a Paragon icon component: `import { Check } from '@edx/paragon/icons';` */
  iconAfter: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
};

WrappedButton.defaultProps = {
  ...Button.defaultProps,
  children: undefined,
  className: undefined,
  iconBefore: undefined,
  iconAfter: undefined,
};

WrappedButton.Deprecated = ButtonDeprecated;

function ButtonGroup(props) {
  return <BaseButtonGroup {...props} />;
}
function ButtonToolbar(props) {
  return <BaseButtonToolbar {...props} />;
}

ButtonGroup.propTypes = {
  /** Specifies element type for this component. */
  as: PropTypes.elementType,
  /** An ARIA role describing the button group. */
  role: PropTypes.string,
  /** Specifies the size for all Buttons in the group. */
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
  /** Display as a button toggle group. */
  toggle: PropTypes.bool,
  /** Specifies if the set of Buttons should appear vertically stacked. */
  vertical: PropTypes.bool,
  /** Overrides underlying component base CSS class name */
  bsPrefix: PropTypes.string,
};

ButtonGroup.defaultProps = {
  as: 'div',
  role: 'group',
  toggle: false,
  vertical: false,
  bsPrefix: 'btn-group',
  size: 'md',
};

ButtonToolbar.propTypes = {
  /** An ARIA role describing the button group. */
  role: PropTypes.string,
  /** Overrides underlying component base CSS class name */
  bsPrefix: PropTypes.string,
};

ButtonToolbar.defaultProps = {
  role: 'toolbar',
  bsPrefix: 'btn-toolbar',
};

export default WrappedButton;
export { ButtonGroup, ButtonToolbar };
