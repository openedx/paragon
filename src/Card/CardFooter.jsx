import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Skeleton from 'react-loading-skeleton';
import CardContext from './CardContext';

const IS_LOADING_HEIGHT_VALUE = 18;

const CardFooter = React.forwardRef(({
  children,
  className,
  isStacked,
  textElement,
  skeletonHeight,
  skeletonWidth,
  orientation,
}, ref) => {
  const { orientation: cardOrientation, isLoading } = useContext(CardContext);
  const footerOrientation = orientation || cardOrientation;
  const wrapperClassName = `pgn__card-footer ${footerOrientation}${isStacked ? '-stacked' : ''}`;
  const textElementClassName = `pgn__card-footer-text ${footerOrientation}${isStacked ? '-stacked' : ''}`;

  if (isLoading) {
    return (
      <div className={classNames(className, wrapperClassName)}>
        <Skeleton
          containerClassName="pgn__card-footer-loader"
          height={skeletonHeight}
          width={skeletonWidth}
        />
      </div>
    );
  }

  return (
    <div className={classNames(className, wrapperClassName)} ref={ref}>
      {textElement && <div className={textElementClassName}>{textElement}</div>}
      {children}
    </div>
  );
});

CardFooter.propTypes = {
  /** Specifies contents of the component. */
  children: PropTypes.node,
  /** Specifies class name to append to the base element. */
  className: PropTypes.string,
  /** Optional node to display near actions. Should be either a plain text or an element containing text (e.g. link). */
  textElement: PropTypes.node,
  /** Specifies whether to use stacked variant. */
  isStacked: PropTypes.bool,
  /** Specifies which orientation to use. This prop will override context value if provided. */
  orientation: PropTypes.oneOf(['horizontal', 'vertical']),
  /** Specifies height of skeleton in loading state. */
  skeletonHeight: PropTypes.number,
  /** Specifies width of skeleton in loading state. */
  skeletonWidth: PropTypes.number,
};

CardFooter.defaultProps = {
  children: null,
  className: undefined,
  textElement: undefined,
  isStacked: false,
  orientation: undefined,
  skeletonHeight: IS_LOADING_HEIGHT_VALUE,
  skeletonWidth: undefined,
};

export default CardFooter;
