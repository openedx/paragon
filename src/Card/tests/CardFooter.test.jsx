import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import Button from '../../Button';
import CardFooter from '../CardFooter';
import CardContext from '../CardContext';

const footerText = 'Sample footer text';
const actions = (
  <>
    <Button>Action 1</Button>
    <Button>Action 2</Button>
  </>
);

// eslint-disable-next-line react/prop-types
function CardFooterWrapper({ orientation = 'vertical', isLoading, ...props }) {
  return (
    <CardContext.Provider value={{ orientation, isLoading }}>
      <CardFooter {...props}>
        {actions}
      </CardFooter>
    </CardContext.Provider>
  );
}

describe('<CardFooter />', () => {
  it('renders vertical orientation without footer text', () => {
    const tree = renderer.create((<CardFooterWrapper />)).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders vertical orientation with footer text', () => {
    const tree = renderer.create((<CardFooterWrapper textElement={footerText} />)).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders stacked vertical orientation without footer text', () => {
    const tree = renderer.create((<CardFooterWrapper isStacked />)).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders stacked vertical orientation with footer text', () => {
    const tree = renderer.create((<CardFooterWrapper textElement={footerText} isStacked />)).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders horizontal orientation without footer text', () => {
    const tree = renderer.create((<CardFooterWrapper orientation="horizontal" />)).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders horizontal orientation with footer text', () => {
    const tree = renderer.create((<CardFooterWrapper orientation="horizontal" textElement={footerText} />)).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders stacked horizontal orientation without footer text', () => {
    const tree = renderer.create((<CardFooterWrapper orientation="horizontal" isStacked />)).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders stacked horizontal orientation with footer text', () => {
    const tree = renderer.create((<CardFooterWrapper orientation="horizontal" isStacked textElement={footerText} />)).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders footer text as element', () => {
    const textElement = <a href="https://example.com">Link text here</a>;
    const { getByText } = render(<CardFooterWrapper textElement={textElement} />);
    const link = getByText('Link text here');

    expect(link).toBeTruthy();
    expect(link.getAttribute('href')).toBe('https://example.com');
  });

  it('renders without loading state', () => {
    const { container } = render(<CardFooterWrapper />);
    expect(container.querySelector('.pgn__card-footer-loader')).toBeNull();
  });

  it('renders with loading state', () => {
    const { container } = render(<CardFooterWrapper isLoading />);
    expect(container.querySelector('.pgn__card-footer-loader')).toBeTruthy();
  });
});
