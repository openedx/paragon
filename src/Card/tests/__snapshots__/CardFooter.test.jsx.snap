// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardFooter /> renders horizontal orientation with footer text 1`] = `
<div
  className="pgn__card-footer horizontal"
>
  <div
    className="pgn__card-footer-text horizontal"
  >
    Sample footer text
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 1
  </button>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 2
  </button>
</div>
`;

exports[`<CardFooter /> renders horizontal orientation without footer text 1`] = `
<div
  className="pgn__card-footer horizontal"
>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 1
  </button>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 2
  </button>
</div>
`;

exports[`<CardFooter /> renders stacked horizontal orientation with footer text 1`] = `
<div
  className="pgn__card-footer horizontal-stacked"
>
  <div
    className="pgn__card-footer-text horizontal-stacked"
  >
    Sample footer text
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 1
  </button>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 2
  </button>
</div>
`;

exports[`<CardFooter /> renders stacked horizontal orientation without footer text 1`] = `
<div
  className="pgn__card-footer horizontal-stacked"
>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 1
  </button>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 2
  </button>
</div>
`;

exports[`<CardFooter /> renders stacked vertical orientation with footer text 1`] = `
<div
  className="pgn__card-footer vertical-stacked"
>
  <div
    className="pgn__card-footer-text vertical-stacked"
  >
    Sample footer text
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 1
  </button>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 2
  </button>
</div>
`;

exports[`<CardFooter /> renders stacked vertical orientation without footer text 1`] = `
<div
  className="pgn__card-footer vertical-stacked"
>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 1
  </button>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 2
  </button>
</div>
`;

exports[`<CardFooter /> renders vertical orientation with footer text 1`] = `
<div
  className="pgn__card-footer vertical"
>
  <div
    className="pgn__card-footer-text vertical"
  >
    Sample footer text
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 1
  </button>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 2
  </button>
</div>
`;

exports[`<CardFooter /> renders vertical orientation without footer text 1`] = `
<div
  className="pgn__card-footer vertical"
>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 1
  </button>
  <button
    className="btn btn-primary"
    disabled={false}
    type="button"
  >
    Action 2
  </button>
</div>
`;
